File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        curr = head
        length = 0

        while curr:
            curr = curr.next
            length += 1
        
        k = k % length
        slow = head
        fast = head
        for i in range(k):
            fast = fast.next

        while fast.next:
            slow = slow.next
            fast = fast.next

        fast.next = head
        newHead = slow.next
        slow.next = None

        return newHead

File: 0061-rotate-list/0061-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    if (!head || !head.next) return head;

    let curr = head, len = 0;
    while (curr) {
        curr = curr.next;
        len++;
    }

    k = k % len;
    let fast = head, slow = head;
    for (let i = 0; i < k; i++) {
        fast = fast.next;
    }

    while(fast.next) {
        slow = slow.next;
        fast = fast.next;
    }

    fast.next = head;
    let newList = slow.next;
    slow.next = null;


    return newList;
};